src/thread_operations.c

```diff
@@ -0,0 +1,83 @@
+#include <FreeRTOS.h>
+#include <semphr.h>
+#include <stdio.h>
+#include <pico/stdlib.h>
+#include <pico/multicore.h>
+#include <pico/cyw43_arch.h>
+#include <stdint.h>
+#include <FreeRTOS.h>
+#include <task.h>
+
+#include "thread_operations.h"
+
+bool side_thread_function(int *counter, SemaphoreHandle_t semaphore)
+{
+    bool semaphore_pass = xSemaphoreTake(semaphore, 500);
+    if(semaphore_pass)
+    {
+        *counter = *counter + 1;
+        xSemaphoreGive(semaphore);
+	    printf("hello world from %s! Count %d\n", "thread", *counter);
+    }
+   
+    return semaphore_pass;
+}
+
+bool main_thread_function(int *counter, int *on, SemaphoreHandle_t semaphore)
+{
+    bool semaphore_pass = xSemaphoreTake(semaphore, 500);
+    if (semaphore_pass)
+    {
+        printf("hello world from %s! Count %d\n", "main", (*counter)++);
+        xSemaphoreGive(semaphore);
+        *on = !(*on);
```

Comment:
This instruction might not need to be in the semaphore if no other thread touches it. In a real project if it might be good practice for either "on" to not be a global variable, or to use a separate semaphore for it - so that the one for "counter" isn't being taken longer than necessary. That being said, it isn't a big deal here.